src/Microsoft.Azure.SignalR/HubHost/ServiceConnection.cs

```diff
@@ -460,20 +460,7 @@ private Task DispatchMessageAsync(ServiceMessage message)
                         try
                         {
                             // Forward the message to the service
-                            if (buffer.IsSingleSegment)
-                            {
-                                await WriteAsync(new ConnectionDataMessage(connection.ConnectionId, buffer.First));
-                            }
-                            else
-                            {
-                                // This is a multi-segmented buffer so just write each chunk
-                                // TODO: Optimize this by doing it all under a single lock
-                                var position = buffer.Start;
-                                while (buffer.TryGet(ref position, out var memory))
-                                {
-                                    await WriteAsync(new ConnectionDataMessage(connection.ConnectionId, memory));
-                                }
-                            }
+                            await WriteAsync(new ConnectionDataMessage(connection.ConnectionId, buffer));
```

Comment:
This is the main change. We write the (potentially multi segment) message directly to the connection output.